### PR TITLE
Support re-iteration in sql connector select, introduce reset method to DataIterator

### PR DESCRIPTION
**Pull request by Manuri**
## Purpose
Support re-iteration in sql connector select, 
Introduce reset method to DataIterator

### CODE REVIEW

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/transaction/mirror_table_xa_transaction_test.bal [hunk 1]
```diff
@@ -191,8 +191,9 @@ function testXATransactionRetry() returns (int, int) {
     } catch (error e) {
     }
     //check whether update action is performed
-    table dt = check testDB1->select("Select COUNT(*) as countval from CustomersTrx where customerId = 4",
-        ResultCount);
+    var temp = testDB1->select("Select COUNT(*) as countval from CustomersTrx where customerId = 4", ResultCount);
+    table dt = check temp;
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/transaction/mirror_table_xa_transaction_test.bal]: Shall we revert this? Any reason for the change?

File: tests/ballerina-test/src/test/resources/test-src/types/table/table_type.bal [hunk 2]
```diff
@@ -410,8 +410,9 @@ function testGetComplexTypes() returns (string, string, string) {
         poolOptions:{maximumPoolSize:1}
     };
 
-    table dt = check testDB->select("SELECT blob_type,clob_type,binary_type from ComplexTypes where row_id = 1", ResultObject
-    );
+    var dtRet = testDB->select("SELECT blob_type,clob_type,binary_type from ComplexTypes where row_id = 1",
+        ResultObject);
+    table dt = check dtRet;
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/types/table/table_type.bal]: Shall we revert this? Any reason for the change?

File: tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal [hunk 3]
```diff
@@ -1003,6 +1009,90 @@ function testComplexTypeRetrieval() returns (string, string, string, string) {
...
+        while (dt.hasNext()) {
+            var rs = check <Employee>dt.getNext();
+            Employee e = {id:rs.id, name:rs.name, address:rs.address};
+            employeeArray3[i] = e;
+            i++;
+        }
+    }
+    catch (error err) {
+        e = err;
+    }
```
Comment (anupama-pathirage) [inline tests/ballerina-test/src/test/resources/test-src/connectors/sql/sql_actions_test.bal]: Formatting issue

**Comment by anupama-pathirage** [comment]
@Manuri Can you please resolve the conflicts?

**Comment by Manuri** [comment]
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/7509

**Pull request merged**